go/multiorch/recovery/healthcheck.go

```diff
@@ -0,0 +1,322 @@
+
+		// Warn if exceeded poll interval
+		pollInterval := re.config.GetPoolerHealthCheckInterval()
+		if totalLatency > pollInterval {
+			if re.poolerPollExceededCounter != nil {
+				re.poolerPollExceededCounter.Add(re.ctx, 1,
+					metric.WithAttributes(
+						attribute.String("database", pooler.MultiPooler.Database),
+						attribute.String("tablegroup", pooler.MultiPooler.TableGroup),
+					))
+			}
+			re.logger.Warn("pollPooler exceeded poll interval",
+				"pooler_id", poolerIDStr,
+				"poll_interval", pollInterval,
+				"actual_duration", totalLatency,
+			)
+		}
+
+		// Record latency histogram
+		if re.pollLatency != nil {
+			re.pollLatency.Record(re.ctx, totalLatency.Seconds(),
+				metric.WithAttributes(
+					attribute.String("database", pooler.MultiPooler.Database),
+					attribute.String("tablegroup", pooler.MultiPooler.TableGroup),
+				))
+		}
+	}()
+
+	// Check cache - prevent redundant checks within poll interval
+	// (Similar to recentDiscoveryOperationKeys in Vitess)
+	// Skip cache check if forceDiscovery is true
+	if !forceDiscovery {
+		if re.existsInCache(poolerIDStr) {
+			// Recently polled, skip
+			re.logger.Debug("skipping poll - recently checked",
+				"pooler_id", poolerIDStr,
+				"poll_interval", re.config.GetPoolerHealthCheckInterval(),
+			)
+			return
+		}
+	}
+
+	// Add to cache
+	re.addToCache(poolerIDStr)
+
+	// Quick check: if up-to-date and valid, skip (unless forceDiscovery)
+	if !forceDiscovery && pooler.IsUpToDate && pooler.IsLastCheckValid {
+		re.logger.Debug("skipping poll - already up to date",
+			"pooler_id", poolerIDStr,
+			"last_check", pooler.LastCheckSuccessful,
+		)
+		return
+	}
+
+	// Log if this is a forced discovery
+	if forceDiscovery {
+		re.logger.Info("force polling pooler",
+			"pooler_id", poolerIDStr,
+			"type", pooler.MultiPooler.Type,
+		)
+	}
+
+	// Increment poll attempts counter
+	if re.pollAttemptsCounter != nil {
+		re.pollAttemptsCounter.Add(re.ctx, 1,
+			metric.WithAttributes(
+				attribute.String("database", pooler.MultiPooler.Database),
+				attribute.String("tablegroup", pooler.MultiPooler.TableGroup),
+			))
+	}
+
+	// Mark attempt timestamp - create new struct to avoid race condition
+	now := time.Now()
+	updated := &store.PoolerHealth{
+		MultiPooler:         pooler.MultiPooler,
+		LastCheckAttempted:  now,
+		LastCheckSuccessful: pooler.LastCheckSuccessful,
+		LastSeen:            pooler.LastSeen,
+		IsUpToDate:          pooler.IsUpToDate,
+		IsLastCheckValid:    pooler.IsLastCheckValid,
+	}
+	re.poolerStore.Set(poolerIDStr, updated)
+	pooler = updated // use updated for rest of function
+
+	// Call appropriate RPC based on pooler type
+	ctx, cancel := context.WithTimeout(re.ctx, 5*time.Second)
+	defer cancel()
+
+	var err error
+	switch pooler.MultiPooler.Type {
+	case clustermetadata.PoolerType_PRIMARY:
+		err = re.pollPrimaryPooler(ctx, poolerID, pooler)
+	case clustermetadata.PoolerType_REPLICA:
+		err = re.pollReplicaPooler(ctx, poolerID, pooler)
```

Comment:
In my opinion, these should not be 2 different functions or RPCs. MultiOrch should ask the multipooler for its information, and it should return information regarding Replica or Primary based on what type it believes itself to be. 
There can be a disparity between what the type of the tablet is according to MultiOrch and what the type of tablet actually is.

Reply:
Cool, this is something that I was debating myself and wondering why it was just one in Vitess. I'll update this in a follow up PR when I actually implement the RPC calls